### PR TITLE
Add apply - a first attempt at adding the apply functionality

### LINKED ISSUE #14: Support arbitrary transformations via `apply` key
Arbitrary functions should be able to be applied to source fields to allow extensibility. adtl will ship with a standard library of functions that are of use to us, but there will be hooks in the library to allow plugging in user defined functions. An example of this could be

```json
{
  "field": "birth_date",
  "apply": "yearsElapsed"
}
```
as an example of an age field which converts the birth date. Either adtl or the user will then need to supply a `yearsElapsed(value)` function which returns the transformed value. All functions should return a valid JSON right hand value (int, float or str).

### PR DESCRIPTION
**Pull request by pipliggins**
fixes #14

### CODE REVIEW

File: adtl/__init__.py [hunk 1]
```diff
@@ -74,6 +76,27 @@ def get_value_unhashed(row: StrDict, rule: Rule) -> Any:
             source_date = get_value(row, rule["source_date"])
             if source_date != target_date:
                 value = datetime.strptime(value, source_date).strftime(target_date)
+        if "apply" in rule and value != "":
+            # apply data transformations.
+            transformation = rule["apply"]["mappingType"]
```
Comment (abhidg) [inline adtl/__init__.py:81]: I think `function` instead of `mappingType` would be better!

File: adtl/__init__.py [hunk 2]
```diff
@@ -74,6 +76,27 @@ def get_value_unhashed(row: StrDict, rule: Rule) -> Any:
...
+            if "apply_params" in rule["apply"]:
+                apply_params = [
+                    row[rule["apply"]["apply_params"][i]]
+                    for i in range(len(rule["apply"]["apply_params"]))
+                ]
+                try:
+                    value = getattr(tf, transformation)(value, *apply_params)
+                except AttributeError:
+                    raise AttributeError(
+                        f"Error using a user-defined transformation: Function {transformation} has not been defined."
```
Comment (abhidg) [inline adtl/__init__.py:91]: These are not user-defined transformations as adtl is supplying them.

File: adtl/transformations.py [hunk 3]
```diff
@@ -0,0 +1,19 @@
+""" Functions which can be applied to source fields, allowing extensibility """
+
+from datetime import datetime
+import pint
+
+
+def isNotNull(value):
+    True if value != None and value != "" else False
```
Comment (abhidg) [inline adtl/transformations.py:8]: needs a return, simpler to use `return value not in [None, ""]`, boolean expressions do not need True / False annotations.

File: adtl/transformations.py [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+
+
+def yearsElapsed(birthdate, currentdate):
+    if birthdate == (None or "") or currentdate == (None or ""):
```
Comment (abhidg) [inline adtl/transformations.py:12]: `birthdate in`, `currentdate in`, the right hand side of equality will resolve to "" so None is not being compared.

File: adtl/__init__.py [hunk 5]
```diff
@@ -74,6 +76,27 @@ def get_value_unhashed(row: StrDict, rule: Rule) -> Any:
             source_date = get_value(row, rule["source_date"])
             if source_date != target_date:
                 value = datetime.strptime(value, source_date).strftime(target_date)
+        if "apply" in rule and value != "":
+            # apply data transformations.
+            transformation = rule["apply"]["mappingType"]
+            if "apply_params" in rule["apply"]:
+                apply_params = [
+                    row[rule["apply"]["apply_params"][i]]
```
Comment (abhidg) [inline adtl/__init__.py:84]: shorten to params?

**Review by abhidg — CHANGES_REQUESTED** [review]
Thanks, approach looks good; some changes required, could you also add tests for adtl/transformations.py?

**Comment by pipliggins** [comment]
Okay, all done!

**Review by abhidg — APPROVED** [review]
Looks good, thanks!

**Pull request merged**